internal/beater/beater.go

```diff
@@ -637,6 +685,37 @@ func (s *Runner) newFinalBatchProcessor(
 	return indexer, indexer.Close, nil
 }
 
+func modelIndexerConfig(
+	opts modelindexer.Config, memLimit float64, logger *logp.Logger,
+) modelindexer.Config {
+	if memLimit < 1 {
+		return opts // use modelindexer defaults
+	}
+	const logMessage = "%s set to %d based on %0.1fgb of memory"
+	opts.EventBufferSize = int(2048 * memLimit)
+	if opts.EventBufferSize >= 61440 {
+		opts.EventBufferSize = 61440
+	}
+	logger.Infof(logMessage,
+		"modelindexer.EventBufferSize", opts.EventBufferSize, memLimit,
+	)
+	if opts.MaxRequests > 0 {
+		return opts
+	}
+	// This formula yields the following max requests for APM Server sized:
+	// 1	2 	4	8	15	30
+	// 10	13	16	22	32	55
+	maxRequests := int(float64(10) + memLimit*1.5)
+	if maxRequests > 60 {
+		maxRequests = 60
+	}
```

Comment:
How did you arrive at these formulae?

If we assumed the max event size (in bytes on the wire) could fit exactly into as many bytes in memory, then isn't a channel size of 2048 for 1GB memory is too much? The default max event size is 300KB. `2048 * 300 * 1024 == 629145600`, which is (substantially) greater than `1024 * 1024 * 1024 == 1073741824`.

What do you think about defining the defaults in terms of:
 - some percentage of available memory
 - maximum event size * (in-memory model objects overhead factor)

?

Reply:
I think that is a big higher than I intended the 1GB instance to hold, however, it would still be within the limit I think?

1,073,741,824 limit in bytes
0,629,145,600 current buffer max usage in bytes (based on 300kb event size). (600MB)

I think it would be safer to to reduce the base queue by 1024 so we end up with up to 300MB of cached events in the queue (And allows some overhead for the in-memory representation).

> How did you arrive at these formulae?

I did some calculations on paper and then adjusted them observing the behavior of benchmarks. The queue didn't have a big impact in throughput to be honest, but It should allow a bit of data to be cached (in practice, it would be less than 1s at maximal throughput, but that's not what the queue is there for).

Modelindexer bulk indexer size = ~10MB (based on some profiles taken that indicated that with 50 available bulk requests, the buffers grew to ~300MB / 50 = ~6MB. per available indexer).

So for a 1GB APM Instance
- Max requests of 128 * 10 * 300kb = 375MB
- Available bulk indexers 11 * 6 = 66MB
- Modelindexer internal queue = 1024 * 300kb = 300MB
- +25% error for other aspects of the server and not consuming all the cgroup memory = 816MB

Reply:
> I think that is a big higher than I intended the 1GB instance to hold, however, it would still be within the limit I think?
>
> 1,073,741,824 limit in bytes
> 0,629,145,600 current buffer max usage in bytes (based on 300kb event size). (600MB)

:man_facepalming: indeed, I was just off by an order of magnitude.